control-plane/api-gateway/common/resources.go

```diff
@@ -401,6 +402,38 @@ func (s *ResourceMap) ExternalFilterExists(filterRef gwv1beta1.LocalObjectRefere
 	return ok
 }
 
+func (s *ResourceMap) AddGatewayPolicy(gatewayPolicy *v1alpha1.GatewayPolicy) *v1alpha1.GatewayPolicy {
+	sectionName := ""
+	if gatewayPolicy.Spec.TargetRef.SectionName != nil {
+		sectionName = string(*gatewayPolicy.Spec.TargetRef.SectionName)
+	}
+	key := api.ResourceReference{
+		Kind:        gatewayPolicy.Spec.TargetRef.Kind,
+		Name:        gatewayPolicy.Spec.TargetRef.Name,
+		SectionName: sectionName,
+		Namespace:   gatewayPolicy.Spec.TargetRef.Namespace,
+	}
+
+	if s.gatewayPolicies == nil {
+		s.gatewayPolicies = make(map[api.ResourceReference]*v1alpha1.GatewayPolicy)
+	}
+
+	s.gatewayPolicies[key] = gatewayPolicy
+
+	return s.gatewayPolicies[key]
+}
+
+func (s *ResourceMap) GetPolicyForGatewayListener(gateway gwv1beta1.Gateway, gatewayListener gwv1beta1.Listener) *v1alpha1.GatewayPolicy {
```

Comment:
Retrieval from a map like this will return an optional boolean. So you can improve the API for this function by returning `(*v1alpha1.GatewayPolicy, bool)`. That way, when you use it, you can have a quick check if the value was found in the map:

```go
if policy, exists := s.GetPolicyForGatewayListener(gateway, gatewayListener); exists {
    [... insert great code here ...]
}
```